mocha/reporter.js

```diff
@@ -0,0 +1,104 @@
+const { v4: uuidv4 } = require('uuid')
+const CI = require('../util/ci')
+const Paths = require('../util/paths')
+const Mocha = require('mocha')
+const Runnable = require('mocha/lib/runnable')
+const uploadTestResults = require('../util/uploadTestResults')
+const failureExpanded = require('../util/failureExpanded')
+
+const {
+  EVENT_RUN_END,
+  EVENT_TEST_BEGIN,
+  EVENT_TEST_END,
+  EVENT_TEST_PENDING,
+} = Mocha.Runner.constants
+
+const {
+  STATE_PASSED,
+  STATE_PENDING,
+  STATE_FAILED,
+} = Runnable.constants
+
+class MochaBuildkiteAnalyticsReporter {
+  constructor(runner) {
+    this._testResults = []
+    this._testEnv = (new CI()).env();
+    this._paths = new Paths({ cwd: process.cwd() }, this._testEnv.location_prefix)
+
+    runner
+      .on(EVENT_TEST_BEGIN, (test) => {
+        this.testStarted(test)
+      })
+      .on(EVENT_TEST_END, (test) => {
+        this.testFinished(test)
+      })
+      .on(EVENT_RUN_END, () => {
+        this.testRunFinished()
+      })
+  }
+
+  testStarted(test) {
+    test.testAnalyticsId = uuidv4()
+    test.startAt = performance.now() / 1000
+  }
+
+  testFinished(test) {
+    const failureReason = test.err !== undefined ? test.err.toString() : undefined
+    const prefixedTestPath = this._paths.prefixTestPath(test.file)
+
+    this._testResults.push({
+      'id': test.testAnalyticsId,
+      'name': test.title,
+      'scope': this.scope(test),
+      'identifier': [this.scope(test), test.title].join(' ').trim(),
+      'file_name': prefixedTestPath,
+      'location': prefixedTestPath,
+      'result': this.analyticsResult(test.state),
+      'failure_reason': failureReason,
+      'failure_expanded': failureExpanded(test.err == undefined ? [] : test.err.multiple),
+      'history': {
+        'section': 'top',
+        'start_at': test.startAt,
+        'end_at': performance.now() / 1000,
```

Comment:
I debated on using performance.now(), but settled on the following. I wasn't entirely sure how accurate the `EVENT_TEST_BEGIN` and `EVENT_TEST_END` hooks were in their timing, so decided to try and have go with the `wallClockStartedAt`. What do you reckon?

```js
    const startTime = new Date(test.wallClockStartedAt).getTime()
    ...
      history: {
        start_at: startTime,
        end_at: startTime + test.duration,
        duration: test.duration / 1000,
      },
```

Reply:
I don't have that property on my test object, I think that may be specific to cypress unfortunatly

Reply:
Ahh fair enough. Performance works on cypress as well so it's the obvious option